packages/flutter/lib/src/widgets/text_formatter.dart

```diff
@@ -0,0 +1,193 @@
+/// IME and not on text under composition (i.e. when 
+/// [TextEditingValue.composing] is collapsed). 
+/// 
+/// Concrete implementations [BlacklistingTextInputFormatter], which removes 
+/// blacklisted characters upon edit commit, and 
+/// [WhitelistingTextInputFormatter], which only allows entries of whitelisted
+/// characters, are provided. 
+/// 
+/// To create custom formatters, extend the [TextInputFormatter] class and 
+/// implement the [formatEditUpdate] method.
+/// 
+/// See also:
+///  
+///  * [EditableText] on which the formatting apply.
+///  * [BlacklistingTextInputFormatter], a provided formatter for blacklisting
+///    characters.
+///  * [WhitelistingTextInputFormatter], a provided formatter for whitelisting
+///    characters.
+abstract class TextInputFormatter {
+  /// Called when text is being typed or cut/copy/pasted in the [EditableText].
+  /// 
+  /// You can override the resulting text based on the previous text value and
+  /// the incoming new text value.
+  /// 
+  /// When formatters are chained, `oldValue` reflects the initial value of
+  /// [TextEditingValue] at the beginning of the chain.
+  TextEditingValue formatEditUpdate(
+    TextEditingValue oldValue, 
+    TextEditingValue newValue
+  );
+
+  /// A shorthand to creating a custom [TextInputFormatter] which formats
+  /// incoming text input changes with the given function.
+  static TextInputFormatter withFunction(
+    TextInputFormatFunction formatFunction
+  ) {
+    return new _SimpleTextInputFormatter(formatFunction);
+  }
+}
+
+/// Function signature expected for creating custom [TextInputFormatter] 
+/// shorthands via [TextInputFormatter.withFunction];
+typedef TextEditingValue TextInputFormatFunction(
+    TextEditingValue oldValue, 
+    TextEditingValue newValue,
+);
+
+/// Wiring for [TextInputFormatter.withFunction].
+class _SimpleTextInputFormatter extends TextInputFormatter {
+  _SimpleTextInputFormatter(this.formatFunction) : 
+    assert(formatFunction != null);
+
+  final TextInputFormatFunction formatFunction;
+  
+  @override
+  TextEditingValue formatEditUpdate(
+    TextEditingValue oldValue, 
+    TextEditingValue newValue
+  ) {
+    return formatEditUpdate(oldValue, newValue);
+  }
+}
+
+/// A [TextInputFormatter] that prevents the insertion of blacklisted 
+/// characters patterns.
+/// 
+/// Instances of blacklisted characters found in the new [TextEditingValue]s 
+/// will be replaced with the [replacementString] which defaults to ``.
+/// 
+/// Since this formatter only removes characters from the text, it attempts to
+/// preserve the existing [TextEditingValue.selection] to values it would now
+/// fall at with the removed characters.
+/// 
+/// See also:
+///  
+///  * [TextInputFormatter].
+///  * [WhitelistingTextInputFormatter].
+class BlacklistingTextInputFormatter extends TextInputFormatter {
+  BlacklistingTextInputFormatter(
+    this.blacklistedPattern, 
+    {this.replacementString: ''}
```

Comment:
put the `{` on the previous line and the `}` on the next line to match similar code elsewhere in the codebase